tools/raml2html/src/Test/ReferenceTester.php

```diff
@@ -0,0 +1,488 @@
+<?php
+
+namespace EzSystems\Raml2Html\Test;
+
+use Symfony\Component\Console\Output\OutputInterface;
+
+class ReferenceTester
+{
+    public const TEST_REFERENCE_ROUTES = 1;
+    public const TEST_CONFIG_ROUTES = 2;
+    public const TEST_ALL_ROUTES = 3;
+
+    public const DEFAULT_FILE_LIST = [
+        'vendor/ibexa/rest/src/bundle/Resources/config/routing.yml',
+        //'vendor/ibexa/commerce-rest/src/bundle/Resources/config/routing.yaml', // Removed as of 4.4
+        // `find $dxpRoot/vendor/ibexa -name "routing_rest.y*ml"`
+        //'vendor/ibexa/admin-ui/src/bundle/Resources/config/routing_rest.yaml',
+        'vendor/ibexa/calendar/src/bundle/Resources/config/routing_rest.yaml',
+        'vendor/ibexa/cart/src/bundle/Resources/config/routing_rest.yaml',
+        'vendor/ibexa/connector-dam/src/bundle/Resources/config/routing_rest.yaml',
+        'vendor/ibexa/personalization/src/bundle/Resources/config/routing_rest.yaml',
+        'vendor/ibexa/product-catalog/src/bundle/Resources/config/routing_rest.yaml',
+        //'vendor/ibexa/scheduler/src/bundle/Resources/config/routing_rest.yaml', // prefixed /api/datebasedpublisher/v1
+        'vendor/ibexa/segmentation/src/bundle/Resources/config/routing_rest.yaml',
+        'vendor/ibexa/taxonomy/src/bundle/Resources/config/routing_rest.yaml',
+    ];
+
+    public const METHOD_LIST = [
+        'OPTIONS',
+        'GET',
+        'HEAD',
+        'POST',
+        'PATCH',
+        'COPY',
+        'MOVE',
+        'SWAP',
+        'PUBLISH',
+        'DELETE',
+    ];
+
+    public $apiUri = '/api/ibexa/v2';
+
+    private const REF_METHOD_NOT_IN_CONF = 'ref_route_method_missing_from_conf';
+    private const CONF_METHOD_NOT_IN_REF = 'conf_route_method_missing_from_ref';
+
+    private $restApiReference;
+    private $dxpRoot;
+
+    private $refRoutes;
+    private $confRoutes;
+
+    /** @var OutputInterface */
+    private $output;
+
+    public function __construct($restApiReference, $dxpRoot, $consolePath = 'bin/console', $routingFiles = null, OutputInterface $output = null)
+    {
+        if (!is_file($restApiReference)) {
+            user_error("$restApiReference doesn't exist or is not a file", E_USER_ERROR);
+            exit(1);
+        }
+        if ('~' === $dxpRoot[0]) {
+            $dxpRoot = trim(shell_exec("echo $dxpRoot"));
+        }
+        if (!is_dir($dxpRoot)) {
+            user_error("$dxpRoot doesn't exist or is not a directory", E_USER_ERROR);
+            exit(2);
+        }
+
+        $this->output = $output;
+
+        $this->restApiReference = $restApiReference;
+        $this->dxpRoot = $dxpRoot;
+        $this->parseApiReference($this->restApiReference);
+        $this->parseRoutes($consolePath, $routingFiles);
+    }
```

Comment:
I've got multiple issues to raise here.

1. `exit` should never be used without reason. It stops the application from working immediately, not giving it a chance to handle an error (be it logging, providing more information to the user, anything). An exception should be thrown instead.
2. Using `shell_exec`. What is happening here exactly? Can't `realpath` be used instead?
3. Constructor should be "light" by default, and do not perform any heavy operations. Any operations that require performing an operation should be moved to a dedicated method.

Regardless of above points, this class is simply too complex. You have a lot of code which requires me to think long what exactly is happening, and there is a high chance I may misunderstand it's purpose. I still think that trying to parse text output from console is a bad idea. All in all, it requires a lot of refactoring at the very least.

Otherwise, I don't see anyone trying to fix anything in this script once it breaks. Maintenance cost is just too high.

Reply:
`realpath` do not solve tilde.
But this is in fact not needed. The shell solve the tilde before passing it to the TestCommand. It's a trace of early development, before it became a Symfony command.

Reply:
Removed in e9807b61524d50e39a0b8005f35848d6f2be6ea0 as the job is done by TestCommand, and if badly done, a good old crash is more verbose.

Reply:
The tool testing the reference against DXP config has been moved to #1901